Add duration parameter in TimeSeasonality

gdeleva opened the pull request:
This PR adds one new parameter to TimeSeasonality: duration.
This determines how long each seasonal period is held constant before moving to the next.

Solves #483

jessegrabowski reviewed (changes requested):
Awesome start! You have exactly the right idea. I think we can be a bit more clever with the math is all.

For the actual implementation, I don't think we can do anything clever -- we'll need to expand the transition matrix to `s * d` and repeat the rows (and also repeat the starting values). The replicates should all be from the same symbolic value, so the user still just provides `s` inputs.

gdeleva commented on pymc_extras/statespace/models/structural.py:
```diff
@@ -1094,14 +1098,24 @@ class TimeSeasonality(Component):
+        \underbrace{\gamma_1, \gamma_1, \ldots, \gamma_1}_{r\ \text{times}},
+        \ldots,
+        \underbrace{\gamma_{s-1}, \gamma_{s-1}, \ldots, \gamma_{s-1}}_{r\ \text{times}},
+        \ldots,
```
@jessegrabowski Thanks. I used the floor operator, though I preferred to split the definition using gamma and \tilde{gamma}. Anyway, we can discuss about it when i formally open a PR.

More importantly, may I ask you if you would expect this line

k_states = season_length - int(self.remove_first_state)

to change to

k_states = season_length*duration - int(self.remove_first_state)?

gdeleva replied:
Thanks for the correction.

gdeleva commented:
@jessegrabowski  This PR may be ready for a review. I added the transition matrices (for both cases depending on the value remove_first_state) in the documentation. Please check if their definitions are aligned with what you had in mind. I updated the class body according to the definitions of these matrices.

gdeleva commented on pymc_extras/statespace/models/structural.py:
```diff
@@ -1094,14 +1098,24 @@ class TimeSeasonality(Component):
+    example. Let :math:`s=4`, :math:`r=1`, and omit the shock term. Define initial conditions :math:`\gamma_0, \gamma_{-1},
     \gamma_{-2}`. The value of the seasonal component for the first 5 timesteps will be:
 
     .. math::
```
Done

AlexAndorra reviewed (changes requested):
Well done @gdeleva , this is looking good! Can you add tests to `test_seasonality.py`, inspiring you from the current tests (feel free to expand them actually, instead of just creating new ones, to avoid redundances). That way we make sure this is actually doing what we want it to do. Also test this when adding to other components, as shown in the corresponding issue

gdeleva commented on pymc_extras/statespace/models/structural/components/seasonality.py:
```diff
@@ -236,8 +337,27 @@ def make_symbolic_graph(self) -> None:
             # all previous states.
-            T = np.eye(k_states, k=-1)
-            T[0, :] = -1
+            zero_d = np.zeros((self.duration, self.duration))
```
@jessegrabowski  Thanks for your review. Would that be okay to build the entire matrix first with numpy and then convert this to a pytensor instead of building pytensor blocks? I am asking because it seems that there is no block method for pytensor.

Also, I did not handle the case remove_first_state=False, because I thought the matrix is essentially determined by k_states = s x d, so the previous code was fine. Please let me know if my understanding is not correct.

gdeleva commented on pymc_extras/statespace/models/structural/components/seasonality.py:
```diff
@@ -146,29 +238,38 @@ def __init__(
         self.provided_state_names = state_names
 
-        k_states = season_length - int(self.remove_first_state)
+        k_states = season_length * duration - int(self.remove_first_state) * duration
```
Thanks. Done

gdeleva commented on pymc_extras/statespace/models/structural/components/seasonality.py:
```diff
@@ -85,10 +148,38 @@ class TimeSeasonality(Component):
     And so on. So for interpretation, the ``season_length - 1`` initial states are, when reversed, the coefficients
     associated with ``state_names[1:]``.
 
+    In the next example, we set :math:`s=3`, :math:`d=2`, ``remove_first_state=True``, and omit the shock term.
```
Thanks. Fixed.

jessegrabowski commented on tests/statespace/models/structural/components/test_seasonality.py:
```diff
@@ -14,62 +14,69 @@
     assert mod.coords["state_season"] == state_names[test_slice]
 
 
+@pytest.mark.parametrize("d", [1, 2, 3])
```
I think just `[1, 3]` is sufficient

gdeleva replied:
Done.

jessegrabowski commented on tests/statespace/models/structural/components/test_seasonality.py:
```diff
@@ -137,20 +146,24 @@ def test_time_seasonality_multiple_observed(rng, remove_first_state):
 
 
-def test_add_two_time_seasonality_different_observed(rng):
+@pytest.mark.parametrize("d1", [1, 2, 3])
+@pytest.mark.parametrize("d2", [1, 2, 3])
```
```suggestion
@pytest.mark.parametrize("d1, d2", [(1, 1),  (1, 3), (3, 1), (3, 3)])
```

Our CI runtime is already pretty hefty

gdeleva replied:
Thanks for the suggestion. Done.

gdeleva replied:
@jessegrabowski No thank you for the support.

jessegrabowski reviewed (changes requested):
This is really great! Just a few final nitpicks to reduce the number of tests, then let's merge it in. Thanks for this

jessegrabowski commented on pymc_extras/statespace/models/structural/components/seasonality.py:
```diff
@@ -146,29 +239,38 @@ def __init__(
             observed_state_names = ["data"]
 
+        if season_length <= 1:
+            raise ValueError(f"season_length must be greater than 1, got {season_length}")
```
I agree with the robot

gdeleva replied:
I also improved the one for season_length by adding the integer check.

AlexAndorra reviewed (approved):
Looking great @gdeleva ! Plus, looks like it's already working with multiple observed time series, so well done 👏
